### PR TITLE
Fix Core Columns butting up against viewport on smaller screens

### PR DESCRIPTION
**Pull request by getdave**
![Screen Shot 2020-03-02 at 15 44 15](https://user-images.githubusercontent.com/444434/75691939-b2a47f80-5c9c-11ea-97e5-8078e9e87dde.png)


On smaller screens, the Core Columns block butts up against the edges of the viewport (unless it has a background set) on both Editor and Frontend.

This PR aims to fix this by introducing left/right spacing around the Block on smaller viewports. This has to be handled differently between Editor and Frontend/Theme due to the way background padding is applied.

See [screenshots below](#screenshots).

My feeling is that without this fix Core Columns looks broken on Themes that don't have specific CSS to add space on left/right. I feel that Core should try to avoid things looking "broken" by default.

I would, however, welcome some guidance and direction from those more experienced with the wider Themes ecosystem.

Also relates to https://github.com/Automattic/wp-calypso/issues/39804.


## How has this been tested?

### Environment
This has been tested using the Gutenberg Starter Theme because it enqueues no Theme specific styles so we see the raw Block styles.

Note that on Twenty Twenty this fix is not needed because the `block-editor-block-list__layout` rules add space on left/right by default. However this is not the case for some Themes.



### Instructions

Switch to "Code View" and paste in the following test Block definitions:

```
<!-- wp:columns {"align":"wide"} -->
<div class="wp-block-columns alignwide"><!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p><strong>Align Wide<br></strong><em>Smoked salmon topped with crème fraîche, dill and capers<br></em>$15.00</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p><strong>Pork Rillettes<br></strong><em>With quick pickle of dried apricots and pistachio crumble<br></em>$14.00</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p><strong>Provençal Vegetable Tart<br></strong><em>With seasonal vegetables<br></em>$12.50</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->

<!-- wp:columns {"backgroundColor":"luminous-vivid-amber","textColor":"very-dark-gray","align":"wide"} -->
<div class="wp-block-columns alignwide has-background has-text-color has-luminous-vivid-amber-background-color has-very-dark-gray-color"><!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p><strong>Align Wide + Background<br></strong><em>Smoked salmon topped with crème fraîche, dill and capers<br></em>$15.00</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p><strong>Pork Rillettes<br></strong><em>With quick pickle of dried apricots and pistachio crumble<br></em>$14.00</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p><strong>Provençal Vegetable Tart<br></strong><em>With seasonal vegetables<br></em>$12.50</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->

<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p><strong>Align Default<br></strong><em>Smoked salmon topped with crème fraîche, dill and capers<br></em>$15.00</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p><strong>Pork Rillettes<br></strong><em>With quick pickle of dried apricots and pistachio crumble<br></em>$14.00</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p><strong>Provençal Vegetable Tart<br></strong><em>With seasonal vegetables<br></em>$12.50</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->

<!-- wp:columns {"align":"full"} -->
<div class="wp-block-columns alignfull"><!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p><strong>Align Full<br></strong><em>Smoked salmon topped with crème fraîche, dill and capers<br></em>$15.00</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p><strong>Pork Rillettes<br></strong><em>With quick pickle of dried apricots and pistachio crumble<br></em>$14.00</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p><strong>Provençal Vegetable Tart<br></strong><em>With seasonal vegetables<br></em>$12.50</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->

<!-- wp:paragraph {"backgroundColor":"luminous-vivid-orange"} -->
<p class="has-background has-luminous-vivid-orange-background-color">Comparr to this please</p>
<!-- /wp:paragraph -->
```

* Install the [Gutenberg Starter Theme](https://github.com/WordPress/gutenberg-starter-theme). 
* On `master`- resize screen to a small viewport 
* Notice how the Columns Block edge butt up against viewport.
* Switch to this PR.
* Resize screen again.
* Notice how the Columns Block edges no longer butt up against viewport.

Do the same on both Editor and when viewing post on Frontend.


## Screenshots <!-- if applicable -->

### Before
<img width="501" alt="Screenshot 2020-03-02 at 15 34 35" src="https://user-images.githubusercontent.com/444434/75691010-a5d35c00-5c9b-11ea-80a3-2e0f3e2ac3e0.png">

---------------
---------------
---------------

### After
<img width="501" alt="Screenshot 2020-03-02 at 15 33 38" src="https://user-images.githubusercontent.com/444434/75691647-40339f80-5c9c-11ea-8d98-6dba71c65d38.png">

## Types of changes
Bug fix (non-breaking change which fixes an issue).

Could possibly break some Themes.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

### CODE REVIEW

**Comment by aduth** [comment]
For prioritization' sake, do you have a sense whether this is a regression of recent changes to the column(s) block(s) in #19515 or #20169?

**Comment by jeryj** [comment]
When I check using the gutenberg starter theme, my editor preview isn't as nicely aligned as your screenshot:
Screen Shot 2020-03-02 at 3 54 14 PM

As a potential help in figuring out why that is, my `master` doesn't even show the yellow/orange-ish background at all. The alignment on the full-width does butt up against the side of the window though:
Screen Shot 2020-03-02 at 3 54 39 PM

File: packages/block-library/src/columns/style.scss [hunk 1]
```diff
@@ -14,6 +14,21 @@
 	}
 }
 
+// Stop Columns butting up against edge of viewport on smaller viewports. Remove
+// on larger viewports where width handles this. Specificity overide required to
+// overide generic `.entry-content *` rule
```
Comment (jeryj) [inline packages/block-library/src/columns/style.scss]: overide should be override

File: packages/block-library/src/columns/style.scss [hunk 2]
```diff
@@ -14,6 +14,21 @@
...
+.wp-block-columns {
+	&.wp-block-columns:not(.has-background) {
+		padding-left: $block-padding;
+		padding-right: $block-padding;
+
+		@include break-xlarge() {
+			padding-left: 0;
+			padding-right: 0;
+		}
```
Comment (jeryj) [inline packages/block-library/src/columns/style.scss]: It looks like if it's an align-full block, that this padding applies, making it butt up against the window on the front-end:
Screen Shot 2020-03-02 at 4 02 18 PM

**Review by jeryj — CHANGES_REQUESTED** [review]
I'm not sure if I tested correctly due to my screenshots being different from yours. Some more eyes on it would be good. If mine ends up not being right I'll remove the change request. Otherwise it's just two tiny comment spelling corrections 👍

**Review by jasmussen — CHANGES_REQUESTED** [review]
Thank you! This improves things, but I don't think it's the right fix, and I think the regression was potentially due to #17877 — sorry about that.

Specifically if I inspect, the margins look correct a great deal of the way down:

Screenshot 2020-03-03 at 09 33 44

Screenshot 2020-03-03 at 09 34 18

So what makes the text butt out like that?

This rule:

Screenshot 2020-03-03 at 09 34 00

That margin-left: -14px (probably -$block-padding in the code) looks like a remnant of when it was necessary to compensate for the left/right margins that have been refactored. 

**And then I git pulled. The above is no longer accurate it seems.** So, master is different now, possibly due to improvements in #20597. 

Now the text only buts out if the columns block is full-wide, and then honestly it seems expected? Because it does so on desktop breakpoints also:

Screenshot 2020-03-03 at 09 42 37

To phrase it differently, there was definitely an issue, but the issue has either changed or even become obsolete overnight. Let me know what you think of master.

**Comment by getdave** [comment]
> For prioritization' sake, do you have a sense whether this is a regression of recent changes to the column(s) block(s) in #19515 or #20169?

@aduth I'm afraid I don't. However, there's a lot of changes to widths and spacing in those PRs so it could be they caused this. I might try testing altering a few of those properties (esp. `margin` changes) to see if that caused it.

I would also say that's it is extremely easy to introduce this kind of regression given the number of variants (Editor, Frontend, different Themes...etc). Not sure how we solve or guard against this...

**Comment by getdave** [comment]
> To phrase it differently, there was definitely an issue, but the issue has either changed or even become obsolete overnight. Let me know what you think of master.

@jasmussen Thanks for this. I've pulled `master` this morning and (testing as per instructions above) it's all fixed!

I'm still concerned this issue could be prevalent [on certain Themes](https://github.com/Automattic/themes) so I'll dig into that, but for now I'm happy to close this out. 🎉 



Screenshot 2020-03-03 at 09 42 38

**Comment by jasmussen** [comment]
Yep. There was a dev-note attached to the margin change, but it is unfortunate that it had to happen. The silver-lining is, however, that the margin math is vastly simpler now.

**Pull request closed without merging**